README
UNDERSTAND HOW TO HIAWATHA MONITOR WORKS
=========================================
Read the information at http://www.hiawatha-webserver.org/howto/monitor to understand what the Hiawatha Monitor is and how it works..


CONFIGURE YOUR WEBSERVER
=========================
Use the following Hiawatha configuration for this website.

UrlToolkit {
	ToolkitID = monitor
	RequestURI isfile Return
	Match ^/(css|images|js)/ Return
	Match ^/(favicon.ico|robots.txt)$ Return
	Match .*\?(.*) Rewrite /index.php?$1
	Match .* Rewrite /index.php
}

VirtualHost {
	Hostname = monitor.domain.com
	WebsiteRoot = /var/www/monitor/public
	AccessLogfile = /var/www/monitor/logfiles/access.log
	ErrorLogfile = /var/www/monitor/logfiles/error.log
	StartFile = index.php
	ExecuteCGI = yes
#	UseFastCGI = PHP5 # Use if you use PHP as a FastCGI daemon
	TimeForCGI = 15
	UseToolkit = monitor
}


CONFIGURE PHP
==============
The Hiawatha Monitor requires the following PHP modules:
	php5-mysql and php5-xsl

Use the following PHP settings:
	cgi.fix_pathinfo = 0 (when using FastCGI PHP), 1 (otherwise)
	cgi.rfc2616_headers = 1
	register_globals = Off
	date.timezone = <your timezone>


CONFIGURE YOUR DATABASE
========================
Change the database settings in 'settings/website.conf' and run the script 'database/setup_database'.


CONFIGURE CRON DAEMON
======================
Use the cron daemon to fetch the information from the webservers:

# Crontab
SHELL=/bin/sh
MAILTO=<account name of which you can read e-mail>
*/5 * * * * /path/to/monitor/website/database/fetch_webserver_logs
3 1 * * *   /path/to/monitor/website/database/delete_old_logs


USING THE HIAWATHA MONITOR
===========================
Login with username 'admin' and password 'monitor' and start adding webservers in the Webserver Administration page. Add "MonitorServer = <IP of monitor server> to the configuration file of your Hiawatha webservers.

Don't forget to change the admin password and to check out the settings in the Settings administration page. When you're done testing, set DEBUG_MODE in settings/website.conf to 'no'.


ChangeLog
monitor (1.3) stable; urgency=low

  * Improved dashboard.
  * Hourly graph added to day overview page.
  * Setup module added.

 -- Hugo Leisink <hugo@leisink.net>  Thu,  3 Sep 2015 14:15:29 +0200

monitor (1.2) stable; urgency=low

  * Banshee framework updated to 5.0.
  * Design made responsive.
  * Database changes:
      CREATE TABLE `organisations` ( `id` int(10) unsigned NOT NULL AUTO_INCREMENT, `name` varchar(50) NOT NULL, PRIMARY KEY (`id`), UNIQUE KEY `name` (`name`), UNIQUE KEY `name_2` (`name`)) ENGINE=InnoDB DEFAULT CHARSET=utf8;
      ALTER TABLE `users` ADD `organisation_id` INT UNSIGNED NOT NULL AFTER `id`, ADD INDEX ( `organisation_id` );
      ALTER TABLE `users` ADD `cert_serial` INT UNSIGNED NULL AFTER `one_time_key`;
      ALTER TABLE `users` CHANGE `password` `password` VARCHAR( 128 ) CHARACTER SET utf8 COLLATE utf8_general_ci NOT NULL;
      Run 'database/private_pages' to update the table 'roles'.
      Run 'database/fix_settings' to add missing settings.
  * Small improvements.

 -- Hugo Leisink <hugo@leisink.net>  Sun, 21 Jun 2015 11:53:02 +0200

monitor (1.1) stable; urgency=low

  * Support for failed login logging.
  * Hostname selection in CMS per user.
  * Improved XML output filter. Add JSON output filter.
  * Reports use median instead of average.
  * Database changes:
      ALTER TABLE `webservers` CHANGE `ssl` `tls` TINYINT(1) NOT NULL;
      ALTER TABLE `host_statistics` ADD `failed_logins` INT UNSIGNED NOT NULL AFTER `exploit_attempts`;
  * Small bugfixes.

 -- Hugo Leisink <hugo@leisink.net>  Sun, 10 May 2015 09:47:41 +0200

monitor (1.0) stable; urgency=low

  * Added script for sending daily reports.
  * Support for Notify My Android notifications.
  * Small improvements.
  * Bugfix: wrong details per day for server statistics.

 -- Hugo Leisink <hugo@leisink.net>  Sun, 12 Oct 2014 09:50:07 +0200

monitor (0.8) stable; urgency=low

  * Access restrictions per webserver added.
  * Added support for CGI errors.
  * Server statistics replaced with connection and bad request graphs.

 -- Hugo Leisink <hugo@leisink.net>  Sat, 31 May 2014 20:07:55 +0200

monitor (0.7) stable; urgency=low

  * Graph layout adepts to amount of monitor days.
  * Added CGI statistics module.
  * Dropped support for request monitoring.
  * Support for Prowl notifications.
  * Improved menu library.
  * Account activity overview added to profile module.
  * Small improvements.

 -- Hugo Leisink <hugo@leisink.net>  Wed, 23 Apr 2014 21:10:07 +0200

monitor (0.6) stable; urgency=low

  * Host statistics module made faster.
  * Pages module added.
  * IP-Country conversion table updated.
  * Small bugfixes.

 -- Hugo Leisink <hugo@leisink.net>  Sun, 13 Dec 2013  8:01:07 +0100

monitor (0.5) stable; urgency=low

  * Renamed actions to events.
  * Error while downloading webserver data logged as event.
  * Generic messages, like high server load, logged as event.
  * Removed unnecessary files and Banshee features.
  * Small improvements and bugfixes.

 -- Hugo Leisink <hugo@leisink.net>  Sat, 31 May 2012 15:54:23 +0200

monitor (0.4) stable; urgency=low

  * Banshee framework updated to version 3.0.
  * Improved User-Agent parsing.
  * Browser version filter in 'Web browsers' and 'Browsers with OSes' page.
  * Detailed information for 'Host statistics' page.
  * Server statistics list for each webserver.
  * Session Manager added.
  * Settings via CMS.
  * Small improvements.

 -- Hugo Leisink <hugo@leisink.net>  Sat,  3 Mar 2012 14:39:44 +0100

monitor (0.3) stable; urgency=low

  * Banshee framework updated to version 2.1.
  * Improved Google referer handling.
  * Count filter for Origins and Referer pages.

 -- Hugo Leisink <hugo@leisink.net>  Mon,  7 Jun 2010 20:47:21 +0200

monitor (0.2) stable; urgency=low

  * Improved database setup script.
  * Banshee bugfixes applied.
  * Bugfix: using port number while fetching log.

 -- Hugo Leisink <hugo@leisink.net>  Mon, 19 Apr 2010 23:30:54 +0200

monitor (0.1) stable; urgency=low

  * Initial release

 -- Hugo Leisink <hugo@leisink.net>  Fri, 12 Feb 2010  9:07:37 +0100


models/setup.php
<?php
	class setup_model extends model {
		private $required_php_extensions = array("libxml", "mysqli", "xsl");

		/* Determine next step
		 */
		public function step_to_take() {
			$missing = $this->missing_php_extensions();
			if (count($missing) > 0) {
				return "php_extensions";
			}

			if ($this->db->connected == false) {
				$db = new MySQLi_connection(DB_HOSTNAME, DB_DATABASE, DB_USERNAME, DB_PASSWORD);
			} else { 
				$db = $this->db;
			}

			if ($db->connected == false) {
				/* No database connection
				 */
				if ((DB_HOSTNAME == "localhost") && (DB_DATABASE == "monitor") && (DB_USERNAME == "monitor") && (DB_PASSWORD == "monitor")) {
					return "db_settings";
				}

				return "create_db";
			}

			$result = $db->execute("show tables like %s", "settings");
			if (count($result) == 0) {
				return "import_sql";
			}

			$settings = new settings($db);
			if ($settings->database_version == null) {
				return "update_db";
			}

			return "done";
		}

		/* Missing PHP extensions
		 */
		public function missing_php_extensions() {
			static $missing = null;

			if ($missing !== null) {
				return $missing;
			}

			$missing = array();
			foreach ($this->required_php_extensions as $extension) {
				if (extension_loaded($extension) == false) {
					array_push($missing, $extension);
				}
			}

			return $missing;
		}

		/* Remove datase related error messages
		 */
		public function remove_database_errors() {
			$errors = explode("\n", rtrim(ob_get_contents()));
			ob_clean();

			foreach ($errors as $error) {
				if (strtolower(substr($error, 0, 14)) != "mysqli_connect") {
					print $error;
				}
			}
		}

		/* Create the MySQL database
		 */
		public function create_database($username, $password) {
			$db = new MySQLi_connection(DB_HOSTNAME, "mysql", $username, $password);

			if ($db->connected == false) {
				$this->output->add_message("Error connecting to database.");
				return false;
			}

			$db->query("begin");

			/* Create database
			 */
			$query = "create database if not exists %S character set utf8";
			if ($db->query($query, DB_DATABASE) == false) {
				$db->query("rollback");
				$this->output->add_message("Error creating database.");
				return false;
			}

			/* Create user
			 */
			$query = "select count(*) as count from user where User=%s";
			if (($users = $db->execute($query, DB_USERNAME)) === false) {
				$db->query("rollback");
				$this->output->add_message("Error checking for user.");
				return false;
			}

			if ($users[0]["count"] == 0) {
				$query = "create user %s@%s identified by %s";
				if ($db->query($query, DB_USERNAME, DB_HOSTNAME, DB_PASSWORD) == false) {
					$db->query("rollback");
					$this->output->add_message("Error creating user.");
					return false;
				}
			} else {
				$login_test = new MySQLi_connection(DB_HOSTNAME, DB_DATABASE, DB_USERNAME, DB_PASSWORD);
				if ($login_test->connected == false) {
					$db->query("rollback");
					$this->output->add_message("Invalid credentials in settings/website.conf.");
					return false;
				}
			}

			/* Set access rights
			 */
			$rights = array(
				"select", "insert", "update", "delete",
				"create", "drop", "alter", "index", "lock tables",
				"create view", "show view");

			$query = "grant ".implode(", ", $rights)." on %S.* to %s@%s";
			if ($db->query($query, DB_DATABASE, DB_USERNAME, DB_HOSTNAME) == false) {
				$db->query("rollback");
				$this->output->add_message("Error setting access rights.");
				return false;
			}

			/* Commit changes
			 */
			$db->query("commit");
			$db->query("flush privileges");
			unset($db);

			return true;
		}

		/* Import SQL script from file
		 */
		public function import_sql() {
			system("mysql -u \"".DB_USERNAME."\" --password=\"".DB_PASSWORD."\" \"".DB_DATABASE."\" < ../database/mysql.sql", $result);
			if ($result != 0) {
				$this->output->add_message("Error while importing database tables.");
				return false;
			}

			$this->db->query("update users set status=%d", USER_STATUS_CHANGEPWD);
			$this->settings->secret_website_code = random_string();

			return true;
		}

		/* Update database
		 */
		public function update_database() {
			$tables = array("cgi_statistics", "host_statistics", "server_statistics");
			foreach ($tables as $table) {
				$this->db->query("alter table %S add %S date not null after %S", $table, "date", "id");
				$this->db->query("alter table %S add %S tinyint unsigned not null after %S", $table, "hour", "date");
				$this->db->query("update %S set %S=date(%S), %S=%d", $table, "date", "timestamp_begin", "hour", 0);
				$this->db->query("alter table %S drop %S", $table, "timestamp_begin");
				$this->db->query("alter table %S drop %S", $table, "timestamp_end");
			}

			$this->db->query("alter table %S add index(%S)", "cgi_statistics", "date");
			$this->db->query("alter table %S add index(%S)", "cgi_statistics", "hour");
			$this->db->query("alter table %S add index(%S)", "host_statistics", "date");
			$this->db->query("alter table %S add index(%S)", "host_statistics", "hour");
			$this->db->query("alter table %S add index(%S)", "server_statistics", "date");
			$this->db->query("alter table %S add index(%S)", "server_statistics", "hour");

			$settings = new settings($this->db);
			$settings->dashboard_threshold_change = 150;
			$settings->dashboard_threshold_value = 5;
			$settings->dashboard_page_refresh = 1;
			$settings->report_alert_high = 300;
			$settings->report_alert_medium = 150;
			$settings->report_history_days = 15;
			$settings->report_skip_normal = false;
			$settings->report_use_median = true;
			$settings->database_version = 103;
		}

		/* Add setting when missing
		 */
		private function ensure_setting($key, $type, $value) {
			if ($this->db->entry("settings", $key, "key") != false) {
				return true;
			}

			$entry = array(
				"key"   => $key,
				"type"  => $type,
				"value" => $value);
			return $this->db->insert("settings", $entry) !== false;
		}

		/* Ensure settings
		 */
		public function ensure_settings() {
			$this->ensure_setting("hiawatha_cache_enabled", "boolean", "false");
			$this->ensure_setting("hiawatha_cache_default_time", "integer", "3600");
			$this->ensure_setting("session_timeout", "integer", "3600");
			$this->ensure_setting("session_persistent", "boolean", "false");
		}
	}
?>
